File: crates/klio-runtime/src/output.rs
use crate::float_fmt;
use std::sync::Arc;

pub trait Output {
    /// Write a string followed by a newline.
    fn writeln(&mut self, s: &str);
    /// Write a string with no trailing newline. Default implementation
    /// stores the partial line and flushes on the next `writeln`.
    fn write(&mut self, s: &str) {
        self.writeln(s);
    }
}

/// One recorded output call. A [`RecordingSink`] logs the exact
/// `write`/`writeln` sequence so it can later be replayed verbatim
/// into any `Output`, byte-for-byte and call-for-call.
#[derive(Clone)]
pub enum OutOp {
    Write(String),
    WriteLn(String),
}

/// `Send` sink that records the exact call sequence instead of
/// formatting eagerly. The root thread and every spawned thread
/// share one of these behind a `Mutex` (via [`SharedOutput`]); when
/// the program finishes, [`RecordingSink::replay_into`] reproduces
/// the calls in order on the caller's real sink. Because the replay
/// is the identical call sequence, a single-threaded program's
/// output is byte-identical to writing the caller's sink directly.
#[derive(Default)]
pub struct RecordingSink {
    pub ops: Vec<OutOp>,
}

impl RecordingSink {
    pub fn replay_into(&mut self, out: &mut dyn Output) {
        for op in self.ops.drain(..) {
            match op {
                OutOp::Write(s) => out.write(&s),
                OutOp::WriteLn(s) => out.writeln(&s),
            }
        }
    }
}

impl Output for RecordingSink {
    fn writeln(&mut self, s: &str) {
        self.ops.push(OutOp::WriteLn(s.to_string()));
    }
    fn write(&mut self, s: &str) {
        self.ops.push(OutOp::Write(s.to_string()));
    }
}

pub struct StdoutOutput;
impl Output for StdoutOutput {
    fn writeln(&mut self, s: &str) {
        println!("{s}");
    }
    fn write(&mut self, s: &str) {
        use std::io::Write;
        let _ = std::io::stdout().write_all(s.as_bytes());
        let _ = std::io::stdout().flush();
    }
}

/// Test helper that captures every line written to it.
#[derive(Default)]
pub struct CaptureOutput {
    pub lines: Vec<String>,
    partial: String,
}
impl Output for CaptureOutput {
    fn writeln(&mut self, s: &str) {
        if self.partial.is_empty() {
            self.lines.push(s.to_string());
        } else {
            let mut joined = std::mem::take(&mut self.partial);
            joined.push_str(s);
            self.lines.push(joined);
        }
    }
    fn write(&mut self, s: &str) {
        self.partial.push_str(s);
    }
}

/// A `Send` output sink shared by every OS thread of one program.
///
/// All threads (the root and every `kotlin.concurrent.thread` child)
/// write through the same `Arc<Mutex<dyn Output + Send>>`, so a
/// `println` is serialized at write granularity. Kotlin does not
/// define an ordering for racy output; the litmus programs gate their
/// prints behind join / `synchronized` so the observable order stays
/// deterministic.
///
/// Single-thread behaviour is byte-identical to writing the inner sink
/// directly: there is exactly one writer, the mutex is uncontended,
/// and `write`/`writeln` forward verbatim in the same program order.
#[derive(Clone, Default)]
pub struct SharedOutput(Arc<std::sync::Mutex<RecordingSink>>);

impl SharedOutput {
    #[must_use] 
    pub fn new() -> Self {
        Self::default()
    }

    /// Replay every recorded call, in order, into the caller's real
    /// sink, then clear the recording. Called once after the run and
    /// every spawned thread have completed.
    pub fn replay_into(&self, out: &mut dyn Output) {
        let mut g = self.0.lock().unwrap_or_else(std::sync::PoisonError::into_inner);
        g.replay_into(out);
    }
}

impl Output for SharedOutput {
    fn writeln(&mut self, s: &str) {
        self.0.lock().unwrap_or_else(std::sync::PoisonError::into_inner).writeln(s);
    }
    fn write(&mut self, s: &str) {
        self.0.lock().unwrap_or_else(std::sync::PoisonError::into_inner).write(s);
    }
}

/// Kotlin-compatible `Double.toString`. Mirrors Java/Kotlin output:
///   * `NaN` literal.
///   * `+/-Infinity` literal.
///   * Integer-valued finite doubles get a `.0` suffix (so `1.0`, not `1`).
///   * Scientific notation kicks in below `1e-3` or at/above `1e7`, with a
///     capital `E` and a `.0` in the mantissa if it's otherwise integer-valued.
#[must_use]
pub fn kotlin_float_to_string(d: f32) -> String {
    float_fmt::float_to_string(d)
}

#[must_use]
pub fn kotlin_double_to_string(d: f64) -> String {
    float_fmt::double_to_string(d)
}

/// Render a single Kotlin `Char` (a UTF-16 code unit) as a `String`. A
/// BMP scalar renders as itself; a lone surrogate renders as the Unicode
/// replacement character, since a Rust `String` cannot hold one.
#[must_use]
pub fn char_unit_to_string(unit: u16) -> String {
    String::from_utf16_lossy(&[unit])
}

/// Append a UTF-16 code unit to a `String`, pairing a pending high
/// surrogate (`prev`) with a following low surrogate into the astral
/// scalar. Returns the new pending high surrogate (the just-appended
/// unit if it is itself an unpaired high surrogate, else `None`). An
/// unpaired surrogate is flushed lossily as U+FFFD. This lets callers
/// fold a `Char` sequence (e.g. a `CharArray`) into a UTF-8 `String`
/// while reconstructing surrogate pairs.
pub fn push_char_unit(out: &mut String, prev: Option<u16>, unit: u16) -> Option<u16> {
    if let Some(hi) = prev {
        if (0xDC00..=0xDFFF).contains(&unit) {
            let c = 0x10000 + ((u32::from(hi) - 0xD800) << 10) + (u32::from(unit) - 0xDC00);
            out.push(char::from_u32(c).unwrap_or('\u{FFFD}'));
            return None;
        }
        out.push('\u{FFFD}'); // unpaired high surrogate
    }
    if (0xD800..=0xDBFF).contains(&unit) {
        return Some(unit); // hold as pending high surrogate
    }
    out.push(char::from_u32(u32::from(unit)).unwrap_or('\u{FFFD}'));
    None
}

/// Fold a sequence of UTF-16 code units into a `String`, reconstructing
/// surrogate pairs (and flushing any trailing unpaired high surrogate).
#[must_use]
pub fn char_units_to_string<I: IntoIterator<Item = u16>>(units: I) -> String {
    let mut out = String::new();
    let mut pending: Option<u16> = None;
    for u in units {
        pending = push_char_unit(&mut out, pending, u);
    }
    if pending.is_some() {
        out.push('\u{FFFD}');
    }
    out
}

File: src/runtime/output.zig
//! Output sink for the interpreter and stdlib, plus the Kotlin-compatible
//! number and char rendering the `Value` display path leans on. `Output` is a
//! `{ctx, vtable}` pair, so every sink presents one interface to intrinsics.

const std = @import("std");
const float_fmt = @import("float_fmt.zig");

/// `writeln` appends a newline; `write` does not.
pub const Output = struct {
    ctx: *anyopaque,
    vtable: *const VTable,

    pub const VTable = struct {
        writeln: *const fn (ctx: *anyopaque, s: []const u8) void,
        write: *const fn (ctx: *anyopaque, s: []const u8) void,
    };

    pub fn writeln(self: Output, s: []const u8) void {
        self.vtable.writeln(self.ctx, s);
    }

    pub fn write(self: Output, s: []const u8) void {
        self.vtable.write(self.ctx, s);
    }
};

pub const OutOp = union(enum) {
    write: []const u8,
    writeln: []const u8,
};

/// Owns the recorded strings; `deinit` frees them.
pub const RecordingSink = struct {
    ops: std.ArrayList(OutOp) = .empty,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) RecordingSink {
        return .{ .ops = .empty, .allocator = allocator };
    }

    pub fn deinit(self: *RecordingSink) void {
        for (self.ops.items) |op| {
            switch (op) {
                .write => |s| self.allocator.free(s),
                .writeln => |s| self.allocator.free(s),
            }
        }
        self.ops.deinit(self.allocator);
    }

    pub fn replayInto(self: *RecordingSink, out: Output) void {
        for (self.ops.items) |op| {
            switch (op) {
                .write => |s| out.write(s),
                .writeln => |s| out.writeln(s),
            }
        }
        for (self.ops.items) |op| {
            switch (op) {
                .write => |s| self.allocator.free(s),
                .writeln => |s| self.allocator.free(s),
            }
        }
        self.ops.clearRetainingCapacity();
    }

    fn vtWriteln(ctx: *anyopaque, s: []const u8) void {
        const self: *RecordingSink = @ptrCast(@alignCast(ctx));
        const owned = self.allocator.dupe(u8, s) catch return;
        self.ops.append(self.allocator, .{ .writeln = owned }) catch {};
    }

    fn vtWrite(ctx: *anyopaque, s: []const u8) void {
        const self: *RecordingSink = @ptrCast(@alignCast(ctx));
        const owned = self.allocator.dupe(u8, s) catch return;
        self.ops.append(self.allocator, .{ .write = owned }) catch {};
    }

    const vtable: Output.VTable = .{ .writeln = vtWriteln, .write = vtWrite };

    pub fn output(self: *RecordingSink) Output {
        return .{ .ctx = self, .vtable = &vtable };
    }
};

pub const StdoutOutput = struct {
    fn vtWriteln(ctx: *anyopaque, s: []const u8) void {
        _ = ctx;
        const w = std.fs.File.stdout().deprecatedWriter();
        w.print("{s}\n", .{s}) catch {};
    }
    fn vtWrite(ctx: *anyopaque, s: []const u8) void {
        _ = ctx;
        const w = std.fs.File.stdout().deprecatedWriter();
        w.writeAll(s) catch {};
    }
    const vtable: Output.VTable = .{ .writeln = vtWriteln, .write = vtWrite };

    pub fn output(self: *StdoutOutput) Output {
        return .{ .ctx = self, .vtable = &vtable };
    }
};

/// `lines` and the pending partial own their bytes; `deinit` frees them.
pub const CaptureOutput = struct {
    lines: std.ArrayList([]const u8) = .empty,
    partial: std.ArrayList(u8) = .empty,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) CaptureOutput {
        return .{ .lines = .empty, .partial = .empty, .allocator = allocator };
    }

    pub fn deinit(self: *CaptureOutput) void {
        for (self.lines.items) |l| self.allocator.free(l);
        self.lines.deinit(self.allocator);
        self.partial.deinit(self.allocator);
    }

    fn vtWriteln(ctx: *anyopaque, s: []const u8) void {
        const self: *CaptureOutput = @ptrCast(@alignCast(ctx));
        if (self.partial.items.len == 0) {
            const owned = self.allocator.dupe(u8, s) catch return;
            self.lines.append(self.allocator, owned) catch {};
        } else {
            self.partial.appendSlice(self.allocator, s) catch {};
            const owned = self.partial.toOwnedSlice(self.allocator) catch return;
            self.lines.append(self.allocator, owned) catch {};
            self.partial = .empty;
        }
    }
    fn vtWrite(ctx: *anyopaque, s: []const u8) void {
        const self: *CaptureOutput = @ptrCast(@alignCast(ctx));
        self.partial.appendSlice(self.allocator, s) catch {};
    }
    const vtable: Output.VTable = .{ .writeln = vtWriteln, .write = vtWrite };

    pub fn output(self: *CaptureOutput) Output {
        return .{ .ctx = self, .vtable = &vtable };
    }
};

pub fn kotlinFloatToString(allocator: std.mem.Allocator, d: f32) ![]u8 {
    return float_fmt.floatToString(allocator, d);
}

pub fn kotlinDoubleToString(allocator: std.mem.Allocator, d: f64) ![]u8 {
    return float_fmt.doubleToString(allocator, d);
}

pub fn charUnitToString(allocator: std.mem.Allocator, unit: u16) ![]u8 {
    return float_fmt.charUnitToString(allocator, unit);
}

/// Pairs a pending high surrogate (`prev`) with a following low surrogate, and
/// answers the new pending high surrogate.
pub fn pushCharUnit(allocator: std.mem.Allocator, out: *std.ArrayList(u8), prev: ?u16, unit: u16) !?u16 {
    if (prev) |hi| {
        if (unit >= 0xDC00 and unit <= 0xDFFF) {
            const c: u32 = 0x10000 + ((@as(u32, hi) - 0xD800) << 10) + (@as(u32, unit) - 0xDC00);
            try appendScalar(allocator, out, c);
            return null;
        }
        // An unpaired high surrogate is a valid Kotlin Char that must
        // round-trip, so keep its WTF-8 form rather than U+FFFD.
        try appendSurrogateUnit(allocator, out, hi);
    }
    if (unit >= 0xD800 and unit <= 0xDBFF) {
        return unit; // hold as pending high surrogate
    }
    if (unit >= 0xDC00 and unit <= 0xDFFF) {
        try appendSurrogateUnit(allocator, out, unit); // lone low surrogate -> WTF-8
        return null;
    }
    try appendScalar(allocator, out, @as(u32, unit));
    return null;
}

fn appendSurrogateUnit(allocator: std.mem.Allocator, out: *std.ArrayList(u8), unit: u16) !void {
    try out.append(allocator, 0xE0 | @as(u8, @intCast(unit >> 12)));
    try out.append(allocator, 0x80 | @as(u8, @intCast((unit >> 6) & 0x3F)));
    try out.append(allocator, 0x80 | @as(u8, @intCast(unit & 0x3F)));
}

fn appendScalar(allocator: std.mem.Allocator, out: *std.ArrayList(u8), c: u32) !void {
    var buf: [4]u8 = undefined;
    const scalar: u21 = if (c <= 0x10FFFF and !(c >= 0xD800 and c <= 0xDFFF))
        @intCast(c)
    else
        0xFFFD;
    const n = std.unicode.utf8Encode(scalar, &buf) catch blk: {
        break :blk std.unicode.utf8Encode(0xFFFD, &buf) catch unreachable;
    };
    try out.appendSlice(allocator, buf[0..n]);
}

pub fn charUnitsToString(allocator: std.mem.Allocator, units: []const u16) ![]u8 {
    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(allocator);
    var pending: ?u16 = null;
    for (units) |u| {
        pending = try pushCharUnit(allocator, &out, pending, u);
    }
    if (pending) |hi| {
        try appendSurrogateUnit(allocator, &out, hi);
    }
    return out.toOwnedSlice(allocator);
}

const testing = std.testing;

test "capture output records full lines" {
    var cap = CaptureOutput.init(testing.allocator);
    defer cap.deinit();
    const out = cap.output();
    out.write("a");
    out.write("b");
    out.writeln("c");
    out.writeln("d");
    try testing.expectEqual(@as(usize, 2), cap.lines.items.len);
    try testing.expectEqualStrings("abc", cap.lines.items[0]);
    try testing.expectEqualStrings("d", cap.lines.items[1]);
}

test "char units to string reconstructs a surrogate pair" {
    const units = [_]u16{ 0xD83D, 0xDE00 }; // U+1F600
    const s = try charUnitsToString(testing.allocator, &units);
    defer testing.allocator.free(s);
    try testing.expectEqualStrings("\u{1F600}", s);
}

test "recording sink replays the exact call sequence" {
    var rec = RecordingSink.init(testing.allocator);
    defer rec.deinit();
    const sink = rec.output();
    sink.write("a");
    sink.write("b");
    sink.writeln("c");
    sink.writeln("d");

    var cap = CaptureOutput.init(testing.allocator);
    defer cap.deinit();
    rec.replayInto(cap.output());

    try testing.expectEqual(@as(usize, 2), cap.lines.items.len);
    try testing.expectEqualStrings("abc", cap.lines.items[0]);
    try testing.expectEqualStrings("d", cap.lines.items[1]);
    try testing.expectEqual(@as(usize, 0), rec.ops.items.len);
}
